File: ios/scanmynet_sdk/Sources/scanmynet_sdk/ScanHostApiImpl.swift
import Foundation
import ScanMyNet

/// Bridges the Pigeon `ScanHostApi` (Dart -> native) onto the RouteThis
/// `ScanMyNet` SDK (`ScanMyNetManager` / `ScanMyNetDelegate`), and forwards the
/// SDK's delegate callbacks back to Dart through `flutterApi`.
///
/// Threading: the SDK drives its services off background queues, so delegate
/// callbacks may arrive on any thread — but the Pigeon channel must be invoked
/// on the main thread. Every `flutterApi` call is therefore funnelled through
/// `onMain`.
///
/// This is the iOS counterpart of the Android `ScanHostApiImpl`. The two SDKs
/// are NOT symmetric (see `pigeons/messages.dart`): iOS has no per-step enum and
/// no "data persisted" debug callback, so `ScanProgress.currentStep`/`nextStep`
/// stay nil (the free-text stage is sent as `label`) and `onDataPersisted` never
/// fires.
final class ScanHostApiImpl: NSObject, ScanHostApi {
  private let flutterApi: ScanFlutterApi
  private var manager: ScanMyNetManager?

  /// The SDK reports `nextScanStarted`/`serviceTriggered` (a label, no percent)
  /// and `scanProgress` (a percent, no label) on separate callbacks. We retain
  /// the last known percent so every `ScanProgress` we emit carries both.
  private var lastPercent: Double = 0

  /// Must match `.schemaVersion` in Dart.
  private static let schemaVersion: Int64 = 3

  init(flutterApi: ScanFlutterApi) {
    self.flutterApi = flutterApi
    super.init()
  }

  // MARK: - ScanHostApi (Dart -> native)

  func configure(config: ScanConfig) throws {
    if let incoming = config.schemaVersion, incoming != Self.schemaVersion {
      throw PigeonError(
        code: "SCHEMA_MISMATCH",
        message: "Pigeon schema mismatch: Dart sent v\(incoming), native expects "
          + "v\(Self.schemaVersion). Codecs are positional — a mismatched pair "
          + "misreads fields silently. Rebuild the plugin's native binaries.",
        details: nil
      )
    }
    lastPercent = 0
    let configuration = ScanMyNetConfiguration(
      apiKey: config.apiKey,
      requestKey: config.requestKey,
      environment: try config.toNativeEnvironment()
    )
    let manager = ScanMyNetManager(configuration: configuration)
    manager.delegate = self
    self.manager = manager
  }

  func startScan() throws {
    guard let manager = manager else {
      throw PigeonError(
        code: "NOT_CONFIGURED",
        message: "configure() must be called before startScan()",
        details: nil
      )
    }
    lastPercent = 0
    manager.start()
  }

  func cancel() throws {
    manager?.cancel()
  }

  // MARK: - Helpers

  private func onMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.async(execute: block)
    }
  }
}

// MARK: - ScanMyNetDelegate (native -> Dart)

extension ScanHostApiImpl: ScanMyNetDelegate {

  func scanStarted() {
    onMain { self.flutterApi.onStarted { _ in } }
  }

  func scanProgress(value: Float) {
    lastPercent = Double(value)
    let progress = ScanProgress(percent: lastPercent)
    onMain { self.flutterApi.onProgress(progress: progress) { _ in } }
  }

  func nextScanStarted(value: String) {
    let progress = ScanProgress(percent: lastPercent, label: value)
    onMain { self.flutterApi.onProgress(progress: progress) { _ in } }
  }

  func serviceTriggered(log: String) {
    let progress = ScanProgress(percent: lastPercent, label: log)
    onMain { self.flutterApi.onProgress(progress: progress) { _ in } }
  }

  /// Required protocol method. The SDK now delivers the payload through
  /// `scanFinished(reportUrl:report:)` below, so this URL-only variant is only
  /// reached on the (unused) default path — emit a result with no report rather
  /// than dropping it.
  func scanFinished(reportUrl: String) {
    let result = ScanResult(reportUrl: reportUrl, status: .success)
    onMain { self.flutterApi.onFinished(result: result) { _ in } }
  }

  /// Carries the decoded report payload. `reportId`/`customerId` are the report
  /// response's siblings of `report`, but the SDK delegate forwards only the
  /// `report` object — so on iOS those two stay nil (they are populated on
  /// Android). Mapping is null-in/null-out via `ReportMapper`.
  func scanFinished(reportUrl: String, report: SMNReportPayload?) {
    let result = ScanResult(
      reportUrl: reportUrl,
      status: .success,
      report: report?.toPigeon()
    )
    onMain { self.flutterApi.onFinished(result: result) { _ in } }
  }

  func scanServiceFailed(type: String, with error: Error) {
    let scanError = ScanError(
      kind: .perStep,
      message: error.localizedDescription,
      serviceType: type
    )
    onMain { self.flutterApi.onError(error: scanError) { _ in } }
  }

  func scanFailed(with error: Error, response: String) {
    let message = response.isEmpty ? error.localizedDescription : response
    let scanError = ScanError(kind: .submission, message: message)
    onMain { self.flutterApi.onError(error: scanError) { _ in } }
  }

  func scanCanceled() {
    onMain { self.flutterApi.onCanceled { _ in } }
  }

  /// Debug-only callback (request/response inspection). Not part of the Pigeon
  /// contract — there is no corresponding `ScanFlutterApi` event — so we ignore it.
  func getRequestData(request: URLRequest?, response: HTTPURLResponse?, error: Error?) {}
}

// MARK: - Enum mapping (Pigeon -> native)

private extension ScanConfig {
  /// Maps the Pigeon environment onto the SDK's `Environment`. The schema
  /// documents production as the default when the host sends nil.
  ///
  /// Unlike Pigeon enums, the native `Environment` carries its URLs as an
  /// associated value, so `.custom` recombines the selector with the separate
  /// `customBaseUrl` / `customFrontendUrl` fields the wire format forces apart.
  func toNativeEnvironment() throws -> Environment {
    switch environment {
    case .staging: return .staging
    case .production: return .production
    // Pigeon `.dev` maps to the SDK's `testing` backend (the dev server).
    case .dev: return .testing
    case .none: return .production
    case .custom:
      // Dart's `configure` rejects this pairing first; this is the backstop for
      // a host that reaches the Pigeon API without going through `ScanmynetSdk`.
      guard let baseUrl = customBaseUrl?.trimmed, !baseUrl.isEmpty else {
        throw PigeonError(
          code: "MISSING_CUSTOM_BASE_URL",
          message: "ScanEnvironment.custom requires customBaseUrl, e.g. "
            + "https://smn.example.com/",
          details: nil
        )
      }
      // A blank frontend URL means "same host serves the viewer" — the SDK
      // falls back to baseUrl, never to our production domain, which would
      // leak the operator's report tokens onto it.
      let frontendUrl = customFrontendUrl?.trimmed
      return .custom(
        baseUrl: baseUrl,
        frontendUrl: (frontendUrl?.isEmpty ?? true) ? nil : frontendUrl
      )
    }
  }
}

private extension String {
  var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

File: android/src/main/kotlin/com/creativeadvtech/scanmynet_sdk/ScanHostApiImpl.kt
package com.creativeadvtech.scanmynet_sdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.creative.tools.network.scan.NetworkScan
import com.creative.tools.network.scan.models.NetworkScanError
import com.creative.tools.network.scan.models.NetworkScanProgress
import com.creative.tools.network.scan.models.NetworkScanResult
import com.creative.tools.network.scan.models.NetworkScanResultStatus
import com.creative.tools.network.scan.models.NetworkScanStep
import com.creative.tools.network.scan.toJson
import com.creative.tools.rest.dto.ReportResponseDto
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.schedulers.Schedulers

/**
 * Bridges the Pigeon [ScanHostApi] (Dart -> native) onto the RouteThis
 * `com.creative.tools.network.scan.NetworkScan` SDK, and forwards the SDK's
 * callbacks back to Dart through [flutterApi].
 *
 * Threading: the SDK's progress/result/debug/error callbacks fire on the scan's
 * RxJava IO scheduler, and the `scan()` `Single` is subscribed on IO — but the
 * Pigeon channel must be invoked on the main thread. Every [flutterApi] call is
 * therefore funnelled through [onMain].
 */
class ScanHostApiImpl(
    private val context: Context,
    private val flutterApi: ScanFlutterApi,
) : ScanHostApi {


    private val mainHandler = Handler(Looper.getMainLooper())
    private val disposables = CompositeDisposable()
    private var networkScan: NetworkScan? = null

    /** Captured from the SDK's ResultCallback to enrich the terminal [ScanResult]. */
    private var lastResult: NetworkScanResult? = null

    /**
     * Frontend (report-viewer) base URL for the configured environment. The
     * backend's `ReportResponseDto.data` carries the report's `verification_token`
     * but not a usable viewer URL, so the terminal [ScanResult.reportUrl] is built
     * here from this base + the token — matching the native app (TestDebugActivity).
     */
    private var frontendBaseUrl: String = DEFAULT_FRONTEND_URL

    override fun configure(config: ScanConfig) {
        val incoming = config.schemaVersion
        if (incoming != null && incoming != SCHEMA_VERSION) {
            throw IllegalStateException(
                "Pigeon schema mismatch: Dart sent v$incoming, native expects " +
                    "v$SCHEMA_VERSION. Codecs are positional — a mismatched pair " +
                    "misreads fields silently. Rebuild the plugin's native binaries.",
            )
        }
        frontendBaseUrl = config.toFrontendUrl()
        networkScan = NetworkScan.builder()
            .context(context)
            .userKey(config.userKey.orEmpty())
            .apiKey(config.apiKey)
            .appName(config.appName.orEmpty())
            .baseUrl(config.toBaseUrl())
            .progressCallback { progress -> onMain { flutterApi.onProgress(progress.toPigeon()) {} } }
            .resultCallback { result -> lastResult = result }
            .debugCallback { report -> onMain { flutterApi.onDataPersisted(ScanReport(report.toJson())) {} } }
            .errorCallback { error -> onMain { flutterApi.onError(error.toPigeon()) {} } }
            .build()
    }

    override fun startScan() {
        val scan = networkScan
            ?: throw IllegalStateException("configure() must be called before startScan()")
        lastResult = null
        onMain { flutterApi.onStarted {} }
        disposables.add(
            scan.scan()
                .subscribeOn(Schedulers.io())
                .subscribe(
                    { response -> onMain { flutterApi.onFinished(response.toPigeon(lastResult)) {} } },
                    { error -> onMain { flutterApi.onError(error.toSubmissionError()) {} } },
                ),
        )
    }

    override fun cancel() {
        // This SDK fork has no native cancel; disposing the subscription stops
        // event delivery as a best-effort cancel (the native tests keep running
        // to completion in the background but their results are dropped).
        disposables.clear()
    }

    /** Releases native resources when the Flutter engine detaches. */
    fun dispose() {
        disposables.clear()
        networkScan = null
        lastResult = null
    }

    private fun onMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) block() else mainHandler.post(block)
    }

    // --- model mappers (native -> Pigeon) ------------------------------------

    private fun NetworkScanProgress.toPigeon() = ScanProgress(
        percent = percent.toDouble() * 100.0, // native percent is a 0..1 fraction
        currentStep = currentStep.toPigeon(),
        nextStep = nextStep.toPigeon(),
        label = null,
        durationMs = duration,
    )

    private fun ReportResponseDto.toPigeon(result: NetworkScanResult?) = ScanResult(
        reportUrl = buildReportUrl(data),
        status = result?.status?.toPigeon() ?: ScanResultStatus.SUCCESS,
        totalDurationMs = result?.duration,
        reportId = report_id,
        customerId = customer_id,
        report = report?.toPigeon(),
    )

    /**
     * Builds the full report-viewer URL Flutter opens as-is. The backend's `data`
     * is a malformed viewer link (e.g. `…mlview-report?verification_token=…`, no
     * `/#/`), so we pull the `verification_token` out and re-assemble it against
     * the environment's [frontendBaseUrl]. `substringAfter` is used instead of
     * `Uri.getQueryParameter` so the token is found whether it sits in the URL's
     * query or its `#/` fragment. If no token is present, `data` is passed through.
     */
    private fun buildReportUrl(data: String): String {
        val marker = "verification_token="
        if (!data.contains(marker)) return data
        val token = data.substringAfter(marker).substringBefore("&")
        return "${frontendBaseUrl}#/view-report/?verification_token=$token"
    }

    private fun NetworkScanError.toPigeon() = ScanError(
        kind = ScanErrorKind.PER_STEP,
        message = error?.message ?: error?.toString() ?: "unknown error",
        step = test.toPigeon(),
        serviceType = null,
    )

    private fun Throwable.toSubmissionError() = ScanError(
        kind = ScanErrorKind.SUBMISSION,
        message = message ?: toString(),
        step = null,
        serviceType = null,
    )

    private fun NetworkScanResultStatus.toPigeon() = when (this) {
        NetworkScanResultStatus.SUCCESS -> ScanResultStatus.SUCCESS
        NetworkScanResultStatus.ERROR_SUBMISSION_FAILED -> ScanResultStatus.SUBMISSION_FAILED
    }

    private fun NetworkScanStep.toPigeon() = when (this) {
        NetworkScanStep.SETUP -> ScanStep.SETUP
        NetworkScanStep.ROUTER_UPNP -> ScanStep.ROUTER_UPNP
        NetworkScanStep.CONNECTIVITY_TEST -> ScanStep.CONNECTIVITY_TEST
        NetworkScanStep.SPEED_TEST -> ScanStep.SPEED_TEST
        NetworkScanStep.PUBLIC_IP_FETCH -> ScanStep.PUBLIC_IP_FETCH
        NetworkScanStep.GPS_LOCATION_FETCH -> ScanStep.GPS_LOCATION_FETCH
        NetworkScanStep.TRACEROUTE -> ScanStep.TRACEROUTE
        NetworkScanStep.WIFI_CONGESTION_TEST -> ScanStep.WIFI_CONGESTION_TEST
        NetworkScanStep.WIFI_DISCOVERY -> ScanStep.WIFI_DISCOVERY
        NetworkScanStep.CONNECTION_QUALITY_TEST -> ScanStep.CONNECTION_QUALITY_TEST
        NetworkScanStep.LAN_DISCOVERY -> ScanStep.LAN_DISCOVERY
        NetworkScanStep.ADDITIONAL_DATA -> ScanStep.ADDITIONAL_DATA
        NetworkScanStep.RESULT_SUBMISSION -> ScanStep.RESULT_SUBMISSION
    }

    companion object {
        /** Must match `.schemaVersion` in Dart. */
        const val SCHEMA_VERSION = 3L
    }
}
